perun-core/src/main/java/cz/metacentrum/perun/core/entry/ServicesManagerEntry.java

```diff
@@ -60,6 +63,227 @@ public ServicesManagerEntry(PerunBl perunBl) {
+					continue; //skip this one because of not authorized action
+				}
+				getServicesManagerBl().unblockAllServicesOnDestination(sess, destination.getId());
+			}
+		}
+	}
+
+	@Override
+	public void unblockAllServicesOnDestination(PerunSession sess, int destinationId) throws PrivilegeException, FacilityNotExistsException, DestinationNotExistsException {
+		Destination destination = getDestinationById(sess, destinationId);
+		List<Facility> destinationFacilities = perunBl.getFacilitiesManager().getFacilitiesByDestination(sess, destination.getDestination());
+
+		boolean isFacilityAdminAtLeastOnce = false;
+		for (Facility facility : destinationFacilities) {
+			if (AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+				isFacilityAdminAtLeastOnce = true;
+			}
+		}
+		if (!isFacilityAdminAtLeastOnce && !AuthzResolver.isAuthorized(sess, Role.PERUNADMIN)) {
+			throw new PrivilegeException(sess, "unblockAllServicesOnDestination");
+		}
+		getServicesManagerBl().unblockAllServicesOnDestination(sess, destinationId);
+	}
+
+	@Override
+	public void unblockServiceOnFacility(PerunSession sess, Service service, Facility facility) throws PrivilegeException {
+		// Authorization
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+			throw new PrivilegeException(sess, "unblockServiceOnFacility");
+		}
+		getServicesManagerBl().unblockServiceOnFacility(sess, service, facility);
+	}
+
+	@Override
+	public void unblockServiceOnDestination(PerunSession sess, Service service, int destinationId) throws PrivilegeException, FacilityNotExistsException, DestinationNotExistsException {
+		Destination destination = getDestinationById(sess, destinationId);
+		List<Facility> destinationFacilities = perunBl.getFacilitiesManager().getFacilitiesByDestination(sess, destination.getDestination());
+
+		boolean isFacilityAdminAtLeastOnce = false;
+		for (Facility facility : destinationFacilities) {
+			if (AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+				isFacilityAdminAtLeastOnce = true;
+			}
+		}
+		if (!isFacilityAdminAtLeastOnce && !AuthzResolver.isAuthorized(sess, Role.PERUNADMIN)) {
+			throw new PrivilegeException(sess, "unblockServiceOnDestination");
+		}
+		getServicesManagerBl().unblockServiceOnDestination(sess, service, destinationId);
+	}
+
+	@Override
+	public boolean forceServicePropagation(PerunSession sess, Facility facility, Service service) throws PrivilegeException {
+		// Authorization
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility) &&
+				!AuthzResolver.isAuthorized(sess, Role.PERUNOBSERVER)) {
+			throw new PrivilegeException(sess, "forceServicePropagation");
+		}
+		return getServicesManagerBl().forceServicePropagation(sess, facility, service);
+	}
+
+	@Override
+	public boolean forceServicePropagation(PerunSession sess, Service service) {
+		return getServicesManagerBl().forceServicePropagation(sess, service);
```

Comment:
We can add generic test, that caller is either ``ENGINE`` or ``FACILITYADMIN``. We could limit it more by checking, whether passed service is assigned to any facility of the facility manager, but its probably not necessary. Please do the same for ``planServicePropagation(Service service)``.

Reply:
This is not yet resolved.